clang/include/clang/CIR/Dialect/IR/CIRTypes.td

```diff
@@ -94,6 +95,82 @@ def SInt16 : SInt<16>;
 def SInt32 : SInt<32>;
 def SInt64 : SInt<64>;
 
+//===----------------------------------------------------------------------===//
+// FloatType
+//===----------------------------------------------------------------------===//
+
+def CIR_FloatType : CIR_Type<"Float", "float",
+    [DeclareTypeInterfaceMethods<DataLayoutTypeInterface>]> {
+  let summary = "CIR floating-point type";
+  let description = [{
+    CIR type that represents C/C++ floating-point types, including `float`, \
+    `double`, and `long double`.
+  }];
+  let parameters = (ins "::mlir::cir::FloatType::Format":$format);
+
+  let builders = [
+    TypeBuilder<(ins CArg<"::mlir::cir::FloatType::Format">:$format)>
+  ];
+
+  let hasCustomAssemblyFormat = 1;
+  let genVerifyDecl = 1;
+
+  let extraClassDeclaration = [{
+    // Floating-point format.
+    enum Format : uint32_t {
+      FF_Binary16,   // IEEE-754 binary16 format
```

Comment:
We should make this more C/C++ like, instead of FF_Binary32, it should be `Float`,  FF_Binary64 should be `Double`, (in the future FF_Binary64E should be `LongDouble`, etc).

Reply:
> (in the future FF_Binary64E should be `LongDouble`, etc).

I disagree.  `long double` representation is a mess.  In different versions of LLVM-based compilers it can have four different formats:

- IEEE 64-bit (`double` in LLVM IR) on Windows, iOS, and MacOS Arm
- X87 80-bit (`x86_fp80`) on Linux x86 and MacOS x86
- IEEE 128-bit (`fp128`) on Linux Arm64
- double-double format (`ppc_fp128`, a very strange and annoying format) on Linux Power

If we want CIR types to correspond to the representation rather than the language type (which has my vote), then none of the types should be named `LongDouble`, because that name is meaningless in the back end.

Reply:
That makes sense indeed. A good part of these enums will be gone by separating the types as suggested by @sitio-couto (at least for the ones that won't change the underlying target type).

@Lancern like I mentioned before, please save all these extra floating point types for next PRs, this one should focus only on float and double. Incremental changes are more likely to get in faster, so I suggest you follow that if you want your changes to land quicker.